src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -945,6 +960,58 @@ internal bool CalculateUseSiteDiagnostic(ref DiagnosticInfo result)
             return false;
         }
 
+#nullable enable
+        protected static UnmanagedCallersOnlyAttributeData DecodeUnmanagedCallersOnlyAttributeData(CSharpAttributeData attribute, Location? location, DiagnosticBag? diagnostics)
+        {
+            Debug.Assert((location == null) == (diagnostics == null));
+            ImmutableHashSet<INamedTypeSymbolInternal>? callingConventionTypes = null;
+            bool isValid = true;
+            if (!attribute.CommonNamedArguments.IsDefaultOrEmpty)
+            {
+                foreach (var (key, value) in attribute.CommonNamedArguments)
+                {
+                    if (key != "CallConvs"
+                        || value.Kind != TypedConstantKind.Array
+                        || value.Values.Any(v => v.Kind != TypedConstantKind.Type))
+                    {
+                        continue;
+                    }
+
+                    if (callingConventionTypes != null)
+                    {
+                        isValid = false;
+                    }
+
+                    foreach (var callConvTypedConstant in value.Values)
+                    {
+                        var builder = PooledHashSet<INamedTypeSymbolInternal>.GetInstance();
+                        Debug.Assert(callConvTypedConstant.Kind == TypedConstantKind.Type);
+                        if (!(callConvTypedConstant.ValueInternal is NamedTypeSymbol callConvType)
+                            || !FunctionPointerTypeSymbol.IsCallingConventionModifier(callConvType))
+                        {
+                            // `{0}` is not a valid calling convention type for 'UnmanagedCallersOnly'.
+                            diagnostics?.Add(ErrorCode.ERR_InvalidUnmanagedCallersOnlyCallConv, location!, callConvTypedConstant.ValueInternal ?? "null");
+                            isValid = false;
+                        }
+                        else
+                        {
+                            _ = builder.Add(callConvType);
+                        }
+
+                        callingConventionTypes = builder.ToImmutableHashSet();
+                        builder.Free();
+                    }
+                }
+            }
+
+            return new UnmanagedCallersOnlyAttributeData(
```

Comment:
>return new UnmanagedCallersOnlyAttributeData( [](start = 12, length = 45)

We should have a singleton instance for `new UnmanagedCallersOnlyAttributeData(ImmutableHashSet<INamedTypeSymbolInternal>.Empty, true)`.